pkg/lightning/log/filter.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package log
+
+import (
+	"strings"
+
+	"go.uber.org/zap/zapcore"
+)
+
+var _ zapcore.Core = (*FilterCore)(nil)
+
+// FilterCore is a zapcore.Core implementation, it filters log by path-qualified
+// package name.
+type FilterCore struct {
+	zapcore.Core
+	filters []string
+}
+
+// NewFilterCore returns a FilterCore.
+//
+// Example, filter TiDB's log, `NewFilterCore(core, "github.com/pingcap/tidb/")`.
+// Note, must set AddCaller() to the logger.
+func NewFilterCore(core zapcore.Core, filteredPackages ...string) *FilterCore {
+	return &FilterCore{
+		Core:    core,
+		filters: filteredPackages,
+	}
+}
+
+// Check overrides wrapper core.Check and adds itself to zapcore.CheckedEntry.
+func (f *FilterCore) Check(entry zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if f.Enabled(entry.Level) {
+		return ce.AddCore(entry, f)
+	}
+	return ce
+}
+
+// Write filters entry by checking if entry's Caller.Function matches filtered
+// package path.
+func (f *FilterCore) Write(entry zapcore.Entry, fields []zapcore.Field) error {
+	for i := range f.filters {
+		// Caller.Function is a package path-qualified function name.
+		if strings.Contains(entry.Caller.Function, f.filters[i]) {
+			return nil
+		}
+	}
+	return f.Core.Write(entry, fields)
+}
```

Comment:
From the [doc's description](https://pkg.go.dev/go.uber.org/zap/zapcore#Core) IMO we should implement the filter logic in `Check()` instead of `Write()`.

`Write` can be kept unimplemented by forwarding to the inner `Core`.

Also you should implement `With()`.

Reply:
I tried `Check`, but I find `Entry.Caller` is not initialized when it is passed `Check()`.  

https://github.com/uber-go/zap/blob/v1.16.0/logger.go#L268-L277

Reply:
Ok. Then can we keep `Check` unimplemented (i.e. forward to `zapcore.Core`)?

Reply:
I tried, we have to override `Check` and add `FilterCore` to `CheckedEntry`, because `Write()` is invoked via `CheckedEntry.core.Write()`.

https://github.com/uber-go/zap/blob/v1.16.0/zapcore/entry.go#L200-L227

Reply:
bruh.